File: java/EncoderDecoder.java
import java.util.Base64;
import java.security.MessageDigest;

public class EncoderDecoder {
    public static String encodeBase64(String inputStr) {
        return Base64.getEncoder().encodeToString(inputStr.getBytes());
    }

    public static String decodeBase64(String inputStr) {
        byte[] decodedBytes = Base64.getDecoder().decode(inputStr);
        return new String(decodedBytes);
    }

    public static String sha256(String inputStr) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-256");
        byte[] hash = digest.digest(inputStr.getBytes());
        StringBuilder hexString = new StringBuilder();
        for (byte b : hash) {
            hexString.append(Integer.toHexString(0xFF & b));
        }
        return hexString.toString();
    }

    public static String sha512(String inputStr) throws Exception {
        MessageDigest digest = MessageDigest.getInstance("SHA-512");
        byte[] hash = digest.digest(inputStr.getBytes());
        StringBuilder hexString = new StringBuilder();
        for (byte b : hash) {
            hexString.append(Integer.toHexString(0xFF & b));
        }
        return hexString.toString();
    }

    public static void main(String[] args) throws Exception {
        java.util.Scanner scanner = new java.util.Scanner(System.in);
        System.out.println("Choose an option:\n 1) Base64 Encode \n2) Base64 Decode \n3) SHA256 \n4) SHA512");
        String choice = scanner.nextLine();
        System.out.println("Enter your message:");
        String inputStr = scanner.nextLine();

        if (choice.equals("1")) {
            System.out.println("Base64 Encoded: " + encodeBase64(inputStr));
        } else if (choice.equals("2")) {
            System.out.println("Base64 Decoded: " + decodeBase64(inputStr));
        } else if (choice.equals("3")) {
            System.out.println("SHA256 Hash: " + sha256(inputStr));
            System.out.println("This cannot be decrypted");
        } else if (choice.equals("4")) {
            System.out.println("SHA512 Hash: " + sha512(inputStr));
            System.out.println("This cannot be decrypted");
        } else {
            System.out.println("Invalid choice");
        }
    }
}

File: scala/src/main/EncoderDecoder/EncoderDecoder.scala
import java.util.Base64
import java.security.MessageDigest

object EncoderDecoder {
  def main(args: Array[String]): Unit = {
    println("Choose encoding method: 1) Base64 Encode 2) Base64 Decode 3) SHA256 4) SHA512")
    val choice = scala.io.StdIn.readLine().toInt
    println("Enter your message:")
    val input = scala.io.StdIn.readLine()

    choice match {
      case 1 =>
        val encoded = Base64.getEncoder.encodeToString(input.getBytes)
        println(s"Encoded Base64: $encoded")
      case 2 =>
        val decoded = new String(Base64.getDecoder.decode(input))
        println(s"Decoded Base64: $decoded")
      case 3 =>
        val sha256 = MessageDigest.getInstance("SHA-256").digest(input.getBytes)
        println(s"SHA256: ${sha256.map("%02x".format(_)).mkString}")
        println("This cannot be decrypted.")
      case 4 =>
        val sha512 = MessageDigest.getInstance("SHA-512").digest(input.getBytes)
        println(s"SHA512: ${sha512.map("%02x".format(_)).mkString}")
        println("This cannot be decrypted.")
      case _ => println("Invalid choice.")
    }
  }
}
